be/src/olap/rowset/segment_v2/column_reader.cpp

```diff
@@ -426,6 +431,20 @@ Status FileColumnIterator::_read_page(const OrdinalPageIndexIterator& iter, Pars
     RETURN_IF_ERROR(_reader->encoding_info()->create_page_decoder(data, options, &page->data_decoder));
     RETURN_IF_ERROR(page->data_decoder->init());
 
+    if (_reader->encoding_info()->encoding() == DICT_ENCODING) {
+        BinaryDictPageDecoder* binary_dict_page_decoder = (BinaryDictPageDecoder*)page->data_decoder;
+        if (binary_dict_page_decoder->is_dict_encoding()) {
+            if (_dict_decoder == nullptr) {
```

Comment:
I think move this new dict_decoder  logic to column_reader because all column iterator use the same dict_decoder and the column reader has to load the dict if the column is dict encoding.

Reply:
for high cardinality column,dict encoding will fallback to plain encoding,its data will contain both encoding, so I think loading dict after data page init and knows the real encoding is better

Reply:
Hi kangping, the reasons for reading dict page inside iterator are

1. Even if a column use dictionary encoding, there could be cases when only non-dict-encoded data pages are seeked. So we lazy read dict page only when necessary
2. `ColumnReader` is owned by Segment and Rowset, being alive even when there is no query. It should retain memory as small as possible.
3. Iterators of the same column won't repeat load the dict page because of page cache.

Reply:
ok, memory is a problems